### PR TITLE
Enable kata-cc image curation on top of mariner 2.0 base image

### PR DESCRIPTION
**Pull request by miz060**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Move kata-cc image curation logic inside AgentBaker so that the Mariner kata-cc release flow can be accomplished all within AKS.

**Which issue(s) this PR fixes**:
Fixes #

**Requirements**:
- [x] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [x] includes documentation
- [x] adds unit tests
- [x] tested upgrade from previous version

**Special notes for your reviewer**:
1. [Kata test pipeline build](https://msazure.visualstudio.com/CloudNativeCompute/_build/results?buildId=75728579&view=logs&j=2ed7bee1-13c6-5ac8-633f-26820f0b06a6&t=25048989-8415-5618-db0f-a93111045e0d) manually verified with AKS node cluster.
2.  All mariner/azure linux build: https://msazure.visualstudio.com/CloudNativeCompute/_build/results?buildId=81213905&view=results

**Release note**:
```
none
```

### CODE REVIEW

File: .pipelines/templates/.builder-release-template.yaml
Comment (miz060) [inline .pipelines/templates/.builder-release-template.yaml:26]: IN PROGRESS: will switch to the latest build once the target pipeline has been stabilized.

File: parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh
Comment (miz060) [inline parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh:43]: IN PROGRESS: cloud-hypervisor-igvm won't be needed with future versions of RPMs.

File: parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh
Comment (miz060) [inline parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh:54]: IN PROGRESS: all blob storage RPMS will be replaced by PMC packages once the new release arrives.

File: vhdbuilder/packer/setupkata.sh
Comment (miz060) [inline vhdbuilder/packer/setupkata.sh:1]: IN PROGRESS: setupkata.sh will be removed once another work stream to enable use of similar script in AKS-RP is completed. This file will be gone before merging.

File: .pipelines/templates/.builder-release-template.yaml
Comment (sprt) [inline .pipelines/templates/.builder-release-template.yaml:33]: nit: If we're specifying a pattern for each file, we might as well list each filename in full, so we know exactly what're we're pulling 🙂

File: parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh
Comment (sprt) [inline parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh:39]: You're missing the file `igvm-measurement.cose`, which is the same `igvm-measurement` but signed.

**Review by sprt — CHANGES_REQUESTED** [review]

File: .pipelines/templates/.builder-release-template.yaml
Comment (dallasd1) [inline .pipelines/templates/.builder-release-template.yaml:33]: The cloud-hypervisor binary is also a temporary artifact until upstreamed by LSG

File: parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh
Comment (dallasd1) [reply parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh:39]: I think we also want the following: igvm.bin, igvm-debug-measurement

File: parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh
Comment (dallasd1) [inline parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh:61]: Once the RPM is published on PMC, we will want to change the default file from `igvm-debug.bin` to `igvm.bin`

File: parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh
Comment (Camelron) [inline parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh:30]: Please also wait for the loader-related changes in my branch to be merged into this branch before merging.

**Review by Camelron — CHANGES_REQUESTED** [review]

File: parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh
Comment (Camelron) [reply parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh:30]: see: https://github.com/Camelron/AgentBaker/commits/cameronbaird/kata-cc-image-curation

File: parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh
Comment (SeanDougherty) [inline parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh:29]: nit: 
It might good to use $OS_VERSION != "1.0", as we will likely want this code to continue to run in future major version releases and this approach could lead to less maintenance.

File: parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh
Comment (SeanDougherty) [inline parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh:40]: Asking to gain understanding:
Similar to line 45, do we need to set permissions on these files?

Comment (SeanDougherty) [reply parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh:40]: Nevermind I see this is handled in packer_source

File: parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh
Comment (SeanDougherty) [inline parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh:45]: Is this chmod required, it seems that this will be handled in packer_source.sh's L130, `cpAndMode $KATA_CLH_SRC $KATA_CLH_DEST 0755`

File: .pipelines/templates/.builder-release-template.yaml
Comment (ms-mahuber) [reply .pipelines/templates/.builder-release-template.yaml:26]: Our new goal was to replicate the UVM build in this pipeline, right?

File: parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh
Comment (ms-mahuber) [inline parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh:53]: what do we need devel for?

File: parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh
Comment (ms-mahuber) [inline parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh:55]: We need the tools to build the UVM. This is probably not happening in this environment. The UVM is built beforehand.

File: parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh
Comment (ms-mahuber) [inline parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh:30]: 1. how about an array or something? anything where we define this list properly
2. for these packages, why do we need to list those explicitly resp., install those: cargo opa parted qemu-img moby-runc python3-devel python3-pip

File: vhdbuilder/scripts/linux/mariner/tool_installs_mariner.sh
Comment (ms-mahuber) [inline vhdbuilder/scripts/linux/mariner/tool_installs_mariner.sh:143]: nit: remove unnecessary comments when emrging

File: vhdbuilder/packer/setupkata.sh
Comment (ms-mahuber) [reply vhdbuilder/packer/setupkata.sh:1]: Just a qq - where will the containerd config setting then go to?

File: vhdbuilder/scripts/linux/mariner/tool_installs_mariner.sh
Comment (ms-mahuber) [inline vhdbuilder/scripts/linux/mariner/tool_installs_mariner.sh:171]: @dallasd1 , @miz060 please revisit - our goal is to build the vanilla kata UVM as well in the UVM build pipeline - at some point we will no longer need this service.

File: vhdbuilder/packer/setupkata.sh
Comment (miz060) [reply vhdbuilder/packer/setupkata.sh:1]: It will be moved to config.toml in AKS-RP.

File: .pipelines/templates/.builder-release-template.yaml
Comment (miz060) [reply .pipelines/templates/.builder-release-template.yaml]: Good suggestion! Updated to list full filenames.

File: parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh
Comment (miz060) [reply parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh]: Got it. Updated with binaries added.

File: parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh
Comment (miz060) [reply parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh:48]: Will do.

File: parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh
Comment (miz060) [reply parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh:30]: Got it.

File: parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh
Comment (miz060) [reply parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh]: Good point! Updated the branch condition as suggested.

File: parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh
Comment (miz060) [reply parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh]: No, it's not. Actually, these `mv` commands are not required neither. Just updated packer_source.sh to copy binaries directly into the confidential-containers directory.

File: parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh
Comment (miz060) [reply parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh]: Got it. kata-containers-cc-tools removed.

File: vhdbuilder/scripts/linux/mariner/tool_installs_mariner.sh
Comment (miz060) [reply vhdbuilder/scripts/linux/mariner/tool_installs_mariner.sh]: Unnecessary comments removed.

File: parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh
Comment (miz060) [reply parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh:41]: Not sure here. @dallasd1 do you know why/if it's needed?

File: parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh
Comment (miz060) [reply parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh]: 1. The general AgentBaker pattern I observed doesn't define list for package dependencies. I think creating one just above the for loop wouldn't be much different. But I agree this part needs a bit of improvement in terms of clarity. Just added a comment to help with that.
2. Good catch! That's a side-effect of installing storage account packages through the RPM tool. Removed non-kata packages here.

File: vhdbuilder/packer/packer_source.sh
Comment (sprt) [inline vhdbuilder/packer/packer_source.sh:130]: On top of this, can we create a file `reference-info-base64` that contains the base64-encoded content of `igvm-measurement.cose`?

**Review by sprt — CHANGES_REQUESTED** [review]

File: vhdbuilder/packer/install-dependencies.sh
Comment (alexeldeib) [inline vhdbuilder/packer/install-dependencies.sh:118]: this check in addition to the v1 check inside the func seems unnecessary? isn't kata v2 only anyway?

File: vhdbuilder/packer/setupkata.sh
Comment (alexeldeib) [inline vhdbuilder/packer/setupkata.sh:12]: is it possible to make this correct by construction like we do in baker.go vs using unstructured text replace?

File: vhdbuilder/scripts/linux/mariner/tool_installs_mariner.sh
Comment (alexeldeib) [inline vhdbuilder/scripts/linux/mariner/tool_installs_mariner.sh:149]: please avoid inline cat when possible - we primarily use it for runtime configuration to avoid adding artifacts to cloud init.

for build time, it should be possible to add file artifacts directly.

File: parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh
Comment (ms-mahuber) [reply parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh:52]: Ace, this will go away soon. This will now be part of separate configs by default in the package. Team, let's highlight such things that will and need to go away for Ace

File: vhdbuilder/packer/packer_source.sh
Comment (sprt) [reply vhdbuilder/packer/packer_source.sh:130]: As discussed offline, instead let's just import the `reference-info-base64` file from the upstream pipeline.

File: vhdbuilder/packer/install-dependencies.sh
Comment (miz060) [reply vhdbuilder/packer/install-dependencies.sh:118]: Yes, for now it's 2.0 only. This is mostly for supporting future major version releases (like Kata v3).

File: vhdbuilder/packer/setupkata.sh
Comment (miz060) [reply vhdbuilder/packer/setupkata.sh:12]: Good suggestion! The current is to actually get rid of this file and use the IsKata() containerd configuration in AKS-RP instead. I'm drafting a design doc and will share it with you as soon as it's ready.

File: .pipelines/templates/.builder-release-template.yaml
Comment (ms-mahuber) [reply .pipelines/templates/.builder-release-template.yaml]: We can resolve this? We are using a variable that we can specify and don't need to set an expicit ID?

File: vhdbuilder/scripts/linux/mariner/tool_installs_mariner.sh
Comment (ms-mahuber) [reply vhdbuilder/scripts/linux/mariner/tool_installs_mariner.sh:158]: @alexeldeib would you like us to turn this into a separate file (e.g. under scripts/linux/mariner) and add it via packer_source.sh to the UVM, or what was your suggested idea?

File: parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh
Comment (miz060) [reply parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh]: Moved inline cat to a separate artifact file.

File: vhdbuilder/scripts/linux/mariner/tool_installs_mariner.sh
Comment (miz060) [reply vhdbuilder/scripts/linux/mariner/tool_installs_mariner.sh]: Moved inline cat to a separate artifact file.

File: vhdbuilder/scripts/linux/mariner/tool_installs_mariner.sh
Comment (Camelron) [inline vhdbuilder/scripts/linux/mariner/tool_installs_mariner.sh:150]: This can be removed **only after** we merge grub2-mkconfig enablement in our core image. There is an additional dependency on a kernel-mshv change which is described here: https://github.com/microsoft/CBL-Mariner/pull/6299#issuecomment-1738246877

Comment (Camelron) [reply vhdbuilder/scripts/linux/mariner/tool_installs_mariner.sh:150]: It is possible we don't merge automatic grub2-mkconfig enablement at all. In that case, there will be a line here calling 

'grub2-mkconfig > /boot/grub2/grub.cfg' 

to put this image on the mkconfig track.

**Review by Camelron — APPROVED** [review]

**Review by Camelron — CHANGES_REQUESTED** [review]
We are going to be in a broken state with the loader unless this is merged:

https://github.com/microsoft/CBL-Mariner/pull/6299

**Comment by Camelron** [comment]
> We are going to be in a broken state with the loader unless this is merged:
> 
> [microsoft/CBL-Mariner#6299](https://github.com/microsoft/CBL-Mariner/pull/6299)

We are now making the switch to mkconfig optional. A separate change can be made after this one to integrate cleanly with a mkconfig-based vhd.

**Review by Camelron — APPROVED** [review]

File: vhdbuilder/scripts/linux/mariner/tool_installs_mariner.sh
Comment (ms-mahuber) [reply vhdbuilder/scripts/linux/mariner/tool_installs_mariner.sh:150]: Future work item when relevant Mariner work is release: https://microsoft.visualstudio.com/OS/_workitems/edit/46713189

**Review by ms-mahuber — APPROVED** [review]

**Review by sprt — APPROVED** [review]

File: parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh
Comment (ms-mahuber) [inline parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh:30]: We should be able to do this now. And if so, we should remove the comment and kernel-uvm-devel from the list

**Review by ms-mahuber — APPROVED** [review]

**Review by dallasd1 — APPROVED** [review]

File: vhdbuilder/packer/vhd-image-builder-mariner.json
Comment (anujmaheshwari1) [inline vhdbuilder/packer/vhd-image-builder-mariner.json:486]: should these changes also be added to the file `AgentBaker/vhdbuilder/packer/vhd-image-builder-mariner-arm64.json`

File: vhdbuilder/packer/packer_source.sh
Comment (anujmaheshwari1) [inline vhdbuilder/packer/packer_source.sh:146]: 755 is a very broad scope of access, other files have 600 or 644, any reason why we are doing 755?

File: vhdbuilder/packer/vhd-image-builder-mariner.json
Comment (miz060) [reply vhdbuilder/packer/vhd-image-builder-mariner.json:486]: No. We don't plan to support aarch64 for kata-cc at this moment.

File: vhdbuilder/packer/packer_source.sh
Comment (miz060) [reply vhdbuilder/packer/packer_source.sh:146]: They needs to be readable and executable for the kata-cc image scenario. I have tested with 644 before that it seems 755 is required.

File: .pipelines/templates/.builder-release-template.yaml
Comment (cameronmeissner) [inline .pipelines/templates/.builder-release-template.yaml:38]: should we only do this when we're building kata distros specifically?

Comment (anujmaheshwari1) [inline .pipelines/templates/.builder-release-template.yaml:38]: do we also want to do this for OS_SKU=AzureLinux? Henry Li made a change recently to onboard AzureLinux and iirc that also has kata flavor.

**Comment by anujmaheshwari1** [comment]
have we tested all VHDs being built with this? I wouldnt expect any issues with Ubuntu SKUs but do we have a test run for all mariner and azurelinux SKUs? if yes, can you please share this in the PR description.

File: .pipelines/templates/.builder-release-template.yaml
Comment (miz060) [reply .pipelines/templates/.builder-release-template.yaml:38]: That would require us to have a separate `vhd-image-builder-mariner-kata.json` as we need to move the downloaded artifact through provisioners and it doesn't take if conditions.

The tradeoff here is to:
- Either have a separate `vhd-image-builder-mariner-kata.json` , which we kinda want to avoid.
- Or Download uvm artifacts for all mariner builds, which we thought to be better as it's ~300MB on the building machine.

Comment (miz060) [reply .pipelines/templates/.builder-release-template.yaml:38]: Good point! That will come with later PRs after preview as we have done multiple rounds of testing with OSSKU=CBLMariner.

OS_SKU=AzureLinux on KataCC needs to be tested more before release.

**Comment by miz060** [comment]
> have we tested all VHDs being built with this? I wouldnt expect any issues with Ubuntu SKUs but do we have a test run for all mariner and azurelinux SKUs? if yes, can you please share this in the PR description.

Yes. The build is a bit old (~2 weeks) so it got deleted. Started and added a new one to the PR description. Will double check before merging.

File: pkg/agent/baker.go
Comment (fangluguomsft) [inline pkg/agent/baker.go:1271]: I only did a high level review. This is what I am looking for. Once this handler is in, I will change AKS code to use this new handler.

The IsKata condition is correct in that Kata-cc still use the same distro/image as Kata.

**Review by fangluguomsft — COMMENTED** [review]
I only did a high level review that is relevant to my code changes. It looks good to me.

**Review by anujmaheshwari1 — APPROVED** [review]
lgtm pending the VHD build is green!

**Review by hbeberman — APPROVED** [review]

**Pull request merged**
